core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1193,7 +1193,14 @@ class DAGScheduler(
             }
 
             if (runningStages.contains(shuffleStage) && shuffleStage.pendingPartitions.isEmpty) {
-              markStageAsFinished(shuffleStage)
+              val noActiveTaskSetManager =
+                taskScheduler.rootPool == null ||
+                  !taskScheduler.rootPool.getSortedTaskSetQueue.exists {
+                    tsm => tsm.stageId == stageId && !tsm.isZombie
+                  }
+              if (shuffleStage.isAvailable || noActiveTaskSetManager) {
+                markStageAsFinished(shuffleStage)
+              }
```

Comment:
I have to admit, though this passes all the tests, this is really confusing to me.  I only somewhat understand why your original version didn't work, and why this should be used instead.  Perhaps some more commenting here would help?  The condition under which you do `markStageAsFinished` seems very broad, so perhaps its worth a comment on the case when you do *not* (and perhaps even a `logInfo` in an `else` branch).  The discrepancy between pendingPartitions and availableOutputs is also surprising -- perhaps that is worth extra comments on `Stage`, on how the meaning of those two are different.